---
title: "Footprinting and post analysis"
output: 
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: false
    code_folding: hide
    theme: "paper"
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE, 
                      message = FALSE, 
                      error = TRUE, 
                      fig.width = 10) 
```

```{r}

library("tidyr")
library("DT")
library("cowplot")
library("tibble")
library("GenomicRanges")
library("biomaRt")
library("ggplot2")
library("plotly")
library("rtracklayer")
library("BSgenome.Hsapiens.UCSC.hg19")
library("rstatix")
library("ggpubr")
library("dplyr")
library("stringr")
library("scales")
library("grid")
library("gridExtra")

options(error=traceback)

```


## Organize initial Homer annotation

```{r}

## Read in called peak information
peaks <- read.delim(file = "All_ALX4_CR_peaks.bed", 
                    header = FALSE)
colnames(peaks) <- c("chr", "start", "end", "name", "score", "strand")

peaks <- peaks %>%
  mutate(PeakSet = case_when(str_detect(name, "SRR24258480") ~ "TWISTposOnly",
                             TRUE ~ "Maintained"))

## Read in exact motif positions
bed <- read.delim(file = "All_ALX4_CR_peaks_motif_TFAPincluded.bed", 
                  header = FALSE, 
                  skip = 1)

colnames(bed) <- c("chr", "start", "end", "Motif", "MotifScore", "Strand")

bed <- mutate(bed, Motif = case_when(Motif == "1-CATCTGKTWTYAATTARN" ~ "Coordinator", 
                                     Motif == "5-TAATYGAATTAG" ~ "Dimer", 
                                     Motif == "2-YAATTA" ~ "Monomer", 
                                     Motif == "1-CAGATG" ~ "Ebox", 
                                     Motif == "4-TTKCCCTGRGGC" ~ "TFAP")) %>%
  arrange(desc(MotifScore))


## Remove mitochondrial DNA
peaks <- peaks %>%
  filter(chr != "chrM")

bed <- bed %>%
  filter(chr != "chrM")

```

## Determine number of possible sites per peak

```{r, fig.height = 5, fig.width = 5}

SiteCounts <- bed %>% group_by(Motif) %>% summarize(Count = n())
SiteCounts$Desc <- "Initial"

colnames(SiteCounts) <- c("Site", "Count", "Desc")

SiteCounts <- relocate(SiteCounts, "Desc", .before = "Site")

## All of my sites are off by 1 if they are on the negative strand, 
## I account for this here

G <- bed %>%
  mutate(start = as.numeric(start), 
         end = as.numeric(end)) %>%
  mutate(start = case_when(Strand == "-" ~ start + 1,
                           TRUE ~ start),
         end = case_when(Strand == "+" ~ end - 1,
                         TRUE ~ end)) %>%
  mutate(start = case_when( Strand == "-" ~ start + 1,
                            TRUE ~ start),
         end = case_when(Strand == "-" ~ end + 1,
                         TRUE ~ end))

G_Sites <- GRanges(seqnames = Rle(G$chr), 
                   ranges = IRanges(start = G$start, 
                                    end = G$end), 
                   name = G$Motif,
                   score = G$MotifScore,
                   strand = G$Strand)


## Combine motif information with peak information

G_peaks <- GRanges(seqnames = Rle(peaks$chr), 
                   ranges = IRanges(start = peaks$start, 
                                    end = peaks$end),
                   names = peaks$name,
                   mcols = peaks %>% select(PeakSet))

OverlapIndices <- findOverlaps(G_Sites, G_peaks, 
                               type = "any", 
                               select = "all") %>%
  as.data.frame()

G$Motif.number <- rownames(G)
peaks$peak.number <- rownames(peaks)

df <- OverlapIndices %>% 
  merge(., G, by.x = "queryHits", by.y = "Motif.number", all.y = TRUE) %>%
  merge(., peaks, by.x = "subjectHits", by.y = "peak.number", all.y = TRUE) %>%
  select(-c(subjectHits, queryHits, chr.x, strand))


colnames(df) <- c("MotifStart", "MotifEnd", "SiteType", "Motif.Score", "strand", "chr", "PeakStart", "PeakEnd", "PeakName", "PeakScore", "PeakSet")
df$Motif_index <- rownames(df)

head(df)

hg19 <- BSgenome.Hsapiens.UCSC.hg19

```


# Motif heatmaps {.tabset}


## Function

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap <- function(df, SiteTypeFilter) {
  
  a <- df %>% filter(SiteType == SiteTypeFilter) %>%
    arrange(Motif.Score)
  
  range = 25
  
  Count <- SiteCounts %>%
    filter(Desc == "Initial") %>%
    filter(Site == SiteTypeFilter) %>%
    .$Count
  
  print(Count)
  
  plot1 <- GRanges(seqnames = Rle(a$chr), 
                   ranges = IRanges(start = a$MotifStart, 
                                    end = a$MotifEnd), 
                   strand = a$strand, 
                   mcols = a %>% select(PeakSet)) %>%
    resize(., fix = "center", width = range * 2) %>%
    getSeq(hg19, .) %>%
    as.data.frame() %>%
    mutate(Motif = rownames(.)) %>%
    separate(col = x, into = c((-range - 1):range) %>% as.character(), sep = "") %>%
    select(-as.character(-range - 1)) %>%
    pivot_longer(!Motif, values_to = "Nucleotide", names_to = "Position") %>%
    mutate(Position = as.numeric(Position)) %>%
    ggplot(aes(x = Position, y = Motif, fill = Nucleotide)) + 
    geom_tile() + 
    scale_fill_manual(values = c("#DD0000", "#00BB00", "#0000EE", "#F8A200"), 
                      breaks = c("T", "A", "C", "G")) + 
    ylab(paste(SiteTypeFilter, " \n(n = ", Count, ")", sep = "")) + 
    scale_x_continuous(limits = c(-range, range), expand = c(0,0)) +
    theme(text = element_text(size = 18), 
          axis.text.y = element_blank(), 
          axis.ticks.y = element_blank(), 
          legend.position = "bottom")
  
  plot2 <- GRanges(seqnames = Rle(a$chr), 
                   ranges = IRanges(start = a$MotifStart, 
                                    end = a$MotifEnd), 
                   strand = a$strand) %>%
    resize(., fix = "center", width = range * 2) %>%
    getSeq(hg19, .) %>%
    as.data.frame() %>%
    mutate(Motif = rownames(.)) %>%
    separate(col = x, into = c((-range-1):range) %>% as.character(), sep = "") %>%
    select(-as.character(-range - 1)) %>%
    pivot_longer(!Motif, values_to = "Nucleotide", names_to = "Position") %>%
    mutate(Position = as.numeric(Position)) %>%
    group_by(Position, Nucleotide) %>%
    summarise(count = n() / as.numeric(Count)) %>%
    ggplot(aes(x = Position, y = count, color = Nucleotide)) + 
    geom_line(linewidth = 1) + 
    scale_x_continuous(limits = c(-range, range), expand = c(0,0)) +
    scale_color_manual(values = c("#DD0000", "#00BB00", "#0000EE", "#F8A200"), 
                       breaks = c("T", "A", "C", "G")) + 
    ylab("Nucleotide\nfraction") + 
    theme_classic() + 
    theme(text = element_text(size = 18),
          legend.position = "none", 
          axis.title.x = element_blank(), 
          axis.ticks.x = element_blank(),
          axis.text.x = element_blank(),
          line = element_blank(),
          axis.ticks.y = element_blank())
  
  plot_grid(plot2, plot1, align = "v", rel_heights = c(2,6), nrow = 2)
  
}

```

## Dimer

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Dimer")

```


## Monomer

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Monomer")

```


## Coordinator

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Coordinator")

```

## Ebox

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Ebox")

```

## TFAP

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "TFAP")

```


These motifs only indicate whether a potential site is within a region of signal, not if that signal is actually bound. 

# Read in sequencing data

## Read in 5' read signal strength

Coverage files were generated using bedtools genomecov tool only from the <120bp regions for CUT&RUN footprinting and all reads for others.

See CoverageCreation.sh script for details. 


## Set regions of interest based on found sites

```{r}

bw_CR_plus <- read.delim("ALX4_CR/SRR24258480.rmdup_plus5_inPeaks.cov", header = FALSE)
bw_CR_plus$strand = "+"
bw_CR_minus <- read.delim("ALX4_CR/SRR24258480.rmdup_minus5_inPeaks.cov", header = FALSE)
bw_CR_minus$strand = "-"

## ASSUMING THAT THIS IS NOT RIGHT INCLUSIVE....

bw_CR <- rbind(bw_CR_plus, bw_CR_minus) %>%
  mutate(index = rownames(.)) %>%
  distinct() %>%
  mutate(V3 = V3 - 1)
colnames(bw_CR) <- c("chr", "start", "end", "depth", "strand", "index")

G_bw_CR <- GRanges(seqnames = Rle(bw_CR$chr),
                   ranges = IRanges(start = bw_CR$start,
                                    end = bw_CR$end),
                   strand = bw_CR$strand,
                   mcols = bw_CR %>% select(depth))

```

## Grab signal strength at motif positions for CUT&RUN 5' ends

```{r}

df_onlySites <- df %>%
  filter( !(is.na(MotifStart)) )

G_Sites <- GRanges(Rle(df_onlySites$chr),
                   IRanges(start = df_onlySites$MotifStart,
                           end = df_onlySites$MotifEnd),
                   names = df_onlySites$Motif_index,
                   strand = df_onlySites$strand,
                   mcols = df_onlySites %>% select(SiteType, MotifStart, MotifEnd, PeakStart, PeakEnd, PeakName, PeakSet))

G_Sites_dispersed <- G_Sites %>%
  resize(width = 202, fix = "center") %>%
  as.data.frame() %>%
  rowwise() %>%
  mutate(binBreaks = ( seq(from = start, to = end, by = 1) %>% as.character() %>% paste(., collapse = ",")) ) %>%
  
  separate(binBreaks, into = c(1:202) %>% as.character(), sep = ",") %>%
  pivot_longer(!c(seqnames:mcols.PeakSet), names_to = "binPosition", values_to = "Coordinate") %>%
  filter( !( is.na(Coordinate) )) %>%
  mutate(Coordinate = as.numeric(Coordinate))

Overlaps <- findOverlaps(GRanges(seqnames = Rle(G_Sites_dispersed$seqnames),
                                 IRanges(start = G_Sites_dispersed$Coordinate,
                                         end = G_Sites_dispersed$Coordinate),
                                 names = G_Sites_dispersed$names,
                                 strand = G_Sites_dispersed$strand,
                                 mcols = G_Sites_dispersed %>% select(mcols.SiteType:mcols.PeakSet)),
                         G_bw_CR) %>%
  as.data.frame()

head(G_Sites_dispersed)

## Grab number of reads for normalization

totalReads <- read.delim(file = "ALX4_CR/SRR24258480_lt120.rmdup.bam_alignment_stats.txt",
                         skip = 6,
                         nrows = 1,
                         header = FALSE) %>%
  .$V3

dfSignal_final <-
  merge(Overlaps, G_Sites_dispersed %>%
          as.data.frame() %>%
          mutate(index = rownames(.)), by.x = "queryHits", by.y = "index") %>%
  merge(., G_bw_CR %>%
          as.data.frame() %>%
          mutate(index = rownames(.)), by.x = "subjectHits", by.y = "index") %>%
  rename(WindowStart = start.x,
         WindowEnd = end.x,
         MotifLength = width.x,
         strand = strand.x,
         seqnames = seqnames.x,
         binStart = start.y,
         binStop = end.y,
         binWidth = width.y,
         CR_signal = mcols.depth,
         PeakSet = mcols.PeakSet,
         name = mcols.SiteType) %>%
  select(-c(subjectHits, queryHits, seqnames.y, strand.y,
            binStart, binStop, binWidth, binPosition)) %>%
  mutate(MotifPosition = Coordinate - mcols.MotifStart - round( (mcols.MotifEnd - mcols.MotifStart) / 2),
         CR_signal = case_when(strand == "-" ~ -CR_signal,
                               TRUE ~ CR_signal)) %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  distinct() %>%
  mutate(CR_signal = CR_signal / ( totalReads / 1000000 ) )

## Ensure number every motif has each position accounted for

dfSignal_final %>%
  ggplot(aes(x = MotifPosition, fill = strand)) +
  geom_bar(width = 1) +
  facet_wrap(~name, nrow = 1) +
  ggtitle("CR signal read position distribution") +
  theme_bw() +
  theme(text = element_text(size = 24),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.y = element_blank(),
        axis.text.x = element_text(angle = 45, 
                                   hjust = 1, 
                                   vjust = 1),
        legend.position = "bottom")

## Read in chromosome sizes
chrom.sizes <- read.delim(file = "hg19.chrom.sizes.txt", 
                          header = FALSE)
colnames(chrom.sizes) <- c("seqnames", "chromSize")


## How many positions do not have full -100 to 100 coverage

print("Number of entries that do not have full coverage in the 100 bp window surrounding motif!")
dfSignal_final %>%
  merge(., chrom.sizes, by = "seqnames") %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  filter(mcols.MotifStart < 0 | mcols.MotifEnd > chromSize )

```

## Add relevant entries to matrix from other sequencing datasets

After each dataset is integrated, I will check to ensure that there are no entries that are NA in the dataset. This will show as empty datasets. 

### Function

```{r}
head(dfSignal_final)

AddtoSignalMatrix <- function(dfSignal,
                              seqType,
                              plusStrand,
                              minusStrand,
                              statsFile,
                              SKIP = 7){
  
  totalReads <- read.delim(file = statsFile,
                           skip = SKIP,
                           nrows = 1,
                           header = FALSE) %>%
    .$V3
  
  bw_plus <- read.delim(plusStrand, header = FALSE)
  bw_plus$strand = "+"
  bw_minus <- read.delim(minusStrand, header = FALSE)
  bw_minus$strand = "-"
  
  bw <- rbind(bw_plus, bw_minus) %>%
    distinct() %>%
    mutate(index = rownames(.)) %>%
    distinct() %>%
    mutate(V3 = V3 - 1)
  colnames(bw) <- c("chr", "start", "end", "depth", "strand", "index")
  
  
  G_bw <- GRanges(seqnames = Rle(bw$chr),
                  ranges = IRanges(start = bw$start,
                                   end = bw$end),
                  strand = bw$strand,
                  mcols = bw %>% select(depth))
  
  
  tmp <- findOverlaps( GRanges(seqnames = Rle(dfSignal$seqnames),
                               ranges = IRanges(start = dfSignal$Coordinate,
                                                end = dfSignal$Coordinate),
                               strand = dfSignal$strand,
                               mcols = dfSignal %>% select(PeakSet)),
                       G_bw) %>%
    as.data.frame() %>%
    distinct() %>%
    merge(., dfSignal %>%
            mutate(index = rownames(.)), by.x = "queryHits", by.y = "index") %>%
    merge(., G_bw %>%
            as.data.frame() %>%
            mutate(index = rownames(.)), by.x = "subjectHits", by.y = "index") %>%
    rename(strand = strand.x,
           seqnames = seqnames.x,
           binStart = start,
           binStop = end,
           binWidth = width,
           NewDepth = mcols.depth) %>%
    select(-c(subjectHits, queryHits, seqnames.y, strand.y)) %>%
    mutate(NewDepth = case_when(strand == "-" ~ -NewDepth,
                                TRUE ~ NewDepth)) %>%
    select(c(seqnames, Coordinate, NewDepth, strand)) %>%
    mutate(NewDepth = NewDepth / ( totalReads / 1000000 ) )
  
  
  dfSignal <- merge(dfSignal, tmp,
                    by = c("Coordinate", "seqnames", "strand"),
                    all.x = TRUE) %>%
    distinct()
  
  
  print("BELOW ENTRIES ARE NA AND NEED ADDRESSED!")
  dfSignal %>% filter(is.na(NewDepth)) %>% print()
  
  return(dfSignal)
  
}

```


## Add datasets to matrix

```{r}

## 5' reads TWIST1 depleted ALX4 CUT&RUN
dfSignal_final <- AddtoSignalMatrix(dfSignal_final,
                                    "CR_TWIST_depleted",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473.rmdup_plus5_inPeaks.cov",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473.rmdup_minus5_inPeaks.cov",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473_lt120.rmdup.bam_alignment_stats.txt",
                                    SKIP = 6) %>%
  rename(TWIST1_depleted = NewDepth)

# 5' reads TWIST1 C&R
dfSignal_final <- AddtoSignalMatrix(dfSignal_final,
                                    "TWIST1_CR",
                                    "TWIST1_CR/SRR24258481.rmdup_plus5_inPeaks.cov",
                                    "TWIST1_CR/SRR24258481.rmdup_minus5_inPeaks.cov",
                                    "TWIST1_CR/SRR24258481_lt120.rmdup.bam_alignment_stats.txt",
                                    SKIP = 6) %>%
  rename(TWIST1_CR = NewDepth)

```


## Initial footprint prior to filter {.tabset}

### Footprinting function

```{r, fig.height = 10, fig.width = 10}

dfSignal_final <- rename(dfSignal_final,
                         Motif_index = names) %>%
  mutate(MotifPosition = Coordinate -
           mcols.MotifStart - round( (mcols.MotifEnd - mcols.MotifStart) / 2))


MotifSortedbyStrength <- dfSignal_final %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  group_by(Motif_index, name) %>% 
  summarize(Avg_CR_sum = sum(CR_signal)) %>% 
  arrange(abs(Avg_CR_sum)) %>% 
  select(Motif_index)

footPrint <- function(dfSignal) {
  
  ID <- dfSignal %>%
    distinct(Motif_index, .keep_all = TRUE) %>%
    select(Motif_index, name) %>%
    mutate(Motif_index = factor(Motif_index, 
                                ordered = TRUE, 
                                #levels = MotifSortedbyStrength
    )) %>%
    group_by(name) %>%
    mutate(ID = row_number())
  
  stichedPlot <- function(dfSignal, NAME) {
    
    if ( NAME == "Coordinator") {
      plot1 <- dfSignal %>%
        filter(name == NAME) %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        group_by(MotifPosition, strand, name) %>%
        summarize(Avg = mean(CR)) %>%
        ggplot(aes(x = MotifPosition, 
                   y = Avg, 
                   fill = strand)) +
        geom_col(show.legend = FALSE, 
                 width = 1) + 
        facet_wrap(~name, nrow = 1) + 
        ylab("RPM") + 
        theme_bw() + 
        ylim(c(-0.04, 0.04)) + 
        scale_fill_manual(values = c("blue", "red")) + 
        scale_x_continuous(expand = c(0,0)) + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.title.x = element_blank(), 
              axis.text.x = element_blank(), 
              axis.ticks.x = element_blank(),
              plot.margin = unit(c(8, 8, 0, 8), "points"), 
              panel.margin = unit(1, "lines"))
      
      plot2 <- dfSignal %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        filter(name == NAME) %>%
        group_by(MotifPosition, Motif_index, name) %>%
        summarize(CR = sum(CR %>% abs())) %>%
        merge(., ID, by = c("Motif_index", "name")) %>%
        ggplot(aes(x = MotifPosition, 
                   y = ID, 
                   fill = CR )) +
        geom_tile() + 
        scale_x_continuous(expand = c(0,0)) + 
        scale_y_discrete(expand = c(0,0)) + 
        scale_fill_gradient(low = "white", 
                            high = "blue", 
                            limits = c(0, .3), 
                            name = "RPM",
                            oob = scales::squish) +
        facet_wrap(~name, nrow = 1)+ 
        ylab("RPM") + 
        xlab("Distance from motif center (bp)") + 
        theme_bw() + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_blank(), 
              axis.ticks.y = element_blank(),
              axis.text.x = element_text(angle = 45, 
                                         hjust = 1, 
                                         vjust = 1), 
              axis.title.x = element_blank(),
              legend.position = "left",
              legend.key.width = unit(1, "cm"),
              legend.direction = "vertical", 
              plot.margin = unit(c(0, 8, 8, 8), "points"), 
              panel.margin = unit(1, "lines")
        )
      
    } else {
      
      plot1 <- dfSignal %>%
        filter(name == NAME) %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        group_by(MotifPosition, strand, name) %>%
        summarize(Avg = mean(CR)) %>%
        ggplot(aes(x = MotifPosition, 
                   y = Avg, 
                   fill = strand)) +
        geom_col(show.legend = FALSE, 
                 width = 1) + 
        facet_wrap(~name, nrow = 1) + 
        ylab("RPM") + 
        theme_bw() + 
        ylim(c(-0.04, 0.04)) + 
        scale_fill_manual(values = c("blue", "red")) + 
        scale_x_continuous(expand = c(0,0)) + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.text.y = element_blank(), 
              axis.ticks.y = element_blank(),
              axis.title.y = element_blank(), 
              axis.title.x = element_blank(), 
              axis.text.x = element_blank(), 
              axis.ticks.x = element_blank(),
              plot.margin = unit(c(8, 8, 0, 8), "points"), 
              panel.margin = unit(1, "lines"))
      
      plot2 <- dfSignal %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        filter(name == NAME) %>%
        group_by(MotifPosition, Motif_index, name) %>%
        summarize(CR = sum(CR %>% abs())) %>%
        merge(., ID, by = c("Motif_index", "name")) %>%
        ggplot(aes(x = MotifPosition, 
                   y = ID, 
                   fill = CR )) +
        geom_tile(show.legend = FALSE) + 
        scale_x_continuous(expand = c(0,0)) + 
        scale_y_discrete(expand = c(0,0)) + 
        scale_fill_gradient(low = "white", 
                            high = "blue", 
                            limits = c(0, .3), 
                            name = "RPM", 
                            oob = scales::squish) +
        facet_wrap(~name, nrow = 1)+ 
        ylab("RPM") + 
        xlab("Distance from motif center (bp)") + 
        theme_bw() + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_blank(), 
              axis.ticks.y = element_blank(),
              axis.text.x = element_text(angle = 45, 
                                         hjust = 1, 
                                         vjust = 1), 
              axis.title.x = element_blank(),
              plot.margin = unit(c(0, 8, 8, 8), "points"), 
              panel.margin = unit(1, "lines")
        )
    }
    
    plot2 <- plot2 + 
      theme(strip.background = element_blank(), strip.text = element_blank())
    
    plot_grid(plot1, plot2, 
              align = "v", 
              rel_heights = c(1,2.75), 
              nrow = 2, 
              axis = "lb")
    
  }
  
  p1 <- stichedPlot(dfSignal, "Coordinator")
  p2 <- stichedPlot(dfSignal, "Dimer")
  p3 <- stichedPlot(dfSignal, "Ebox")
  p4 <- stichedPlot(dfSignal, "Monomer")
  p5 <- stichedPlot(dfSignal, "TFAP")
  
  plot <- plot_grid(p1, p2, p3, p4, p5,
                    align = "h", 
                    rel_widths = c(1.52,1,1,1,1), 
                    nrow = 1)
  
  x.grob <- textGrob("Distance from motif center (bp)", 
                     gp=gpar(fontsize=24))
  
  return(grid.arrange(arrangeGrob(plot, bottom = x.grob)))
  
}


```

### ALX4 CUT&RUN

```{r, fig.height = 8, fig.width = 10}

footPrint(dfSignal_final %>%
            mutate(CR = CR_signal) )

```

### TWIST1 CUT&RUN

```{r, fig.height = 8, fig.width = 10}

footPrint(dfSignal_final %>%
            mutate(CR = TWIST1_CR)) 


```

### ALX4 CUT&RUN in TWIST1 depleted conditions

```{r, fig.height = 8, fig.width = 10}

footPrint(dfSignal_final  %>%
            mutate(CR = TWIST1_depleted)) 

```



## Select sites based on footprinting {.tabset}

### Function

```{r}

footprintFilter <- function(dfSignal_final) {
  
  SiteCounts <- bed %>% group_by(Motif) %>% summarize(Count = n())
  SiteCounts$Desc <- "Initial"
  
  colnames(SiteCounts) <- c("Site", "Count", "Desc")
  
  SiteCounts <- relocate(SiteCounts, "Desc", .before = "Site")
  
  dimerIndex <- dfSignal_final %>% 
    filter(name == "Dimer") %>%
    mutate(Position = case_when(MotifPosition %in% c(-6:6) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-46:-7), c(7:46)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1)
  
  dimerIndexDF <- dfSignal_final %>% 
    filter(name == "Dimer") %>%
    mutate(Position = case_when(MotifPosition %in% c(-6:6) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-46:-7), c(7:46)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) )
  
  dimerIndex$Site <- "Dimer"
  
  monomerIndex <- dfSignal_final %>% 
    filter(name == "Monomer") %>%
    mutate(Position = case_when(MotifPosition %in% c(-3:3) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-43:-4), c(4:43)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1) 
  
  monomerIndex$Site <- "Monomer"
  
  coordinatorIndex <- dfSignal_final %>% 
    filter(name == "Coordinator") %>%
    mutate(Position = case_when(MotifPosition %in% c(-10:10) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-50:-11), c(11:50)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1) 
  
  coordinatorIndex$Site <- "Coordinator"
  
  EboxIndex <- dfSignal_final %>% 
    filter(name == "Ebox") %>%
    mutate(Position = case_when(MotifPosition %in% c(-3:3) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-43:-4), c(4:43)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1)
  
  EboxIndex$Site <- "Ebox"
  
  TFAPIndex <- dfSignal_final %>% 
    filter(name == "TFAP") %>%
    mutate(Position = case_when(MotifPosition %in% c(-6:6) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-46:-7), c(7:46)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1)
  
  TFAPIndex$Site <- "TFAP"
  
  SiteCounts <- rbind(SiteCounts, 
                      c("MNAsefilter", "Monomer", dim(monomerIndex)[1]),
                      c("MNAsefilter", "Dimer", dim(dimerIndex)[1]), 
                      c("MNAsefilter", "Coordinator", dim(coordinatorIndex)[1]), 
                      c("MNAsefilter", "Ebox", dim(EboxIndex)[1]), 
                      c("MNAsefilter", "TFAP", dim(TFAPIndex)[1]))
  
  
  IndexDF <- rbind(dimerIndex, monomerIndex, coordinatorIndex, EboxIndex, TFAPIndex)
  
  
  tmp <- df %>% 
    distinct(Motif_index, .keep_all = TRUE) %>%
    merge(., IndexDF, by = "Motif_index") %>% 
    filter( !is.na(Site)) %>% 
    arrange(Thres %>% desc())
  
  Gtmp <- GRanges(seqnames = Rle(tmp$chr), 
                  ranges = IRanges(start = tmp$MotifStart, 
                                   end = tmp$MotifEnd), 
                  name = tmp$Motif_index,
                  strand = tmp$strand, 
                  mcols = tmp %>% select(MotifFlank:Site))
  
  G_Ebox <- Gtmp[mcols(Gtmp)$mcols.Site == "Ebox"]
  G_mon <- Gtmp[mcols(Gtmp)$mcols.Site == "Monomer"]
  G_dimer <- Gtmp[mcols(Gtmp)$mcols.Site == "Dimer"]
  G_coord <- Gtmp[mcols(Gtmp)$mcols.Site == "Coordinator"]
  G_TFAP <- Gtmp[mcols(Gtmp)$mcols.Site == "TFAP"]
  
  
  ## Filter by self overlap
  
  G_mon <- G_mon[unique(findOverlaps(G_mon, 
                                     ignore.strand = TRUE,
                                     type = "any", 
                                     select = "first"))]
  
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Monomer", length(G_mon)))
  
  
  G_coord <- G_coord[c(unique(findOverlaps(G_coord,
                                         ignore.strand = TRUE, 
                                         type = "any", 
                                         select = "first")), 2939)]
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Coordinator", length(G_coord)))
  
  G_dimer <- G_dimer[unique(findOverlaps(G_dimer, 
                                         ignore.strand = TRUE,
                                         type = "any", 
                                         select = "first"))]
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Dimer", length(G_dimer)))
  
  G_Ebox <- G_Ebox[unique(findOverlaps(G_Ebox, 
                                       ignore.strand = TRUE,
                                       type = "any", 
                                       select = "first"))]
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Ebox", length(G_Ebox)))
  
  G_TFAP <- G_TFAP[unique(findOverlaps(G_TFAP, 
                                       ignore.strand = TRUE,
                                       type = "any", 
                                       select = "first"))]
  
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "TFAP", length(G_TFAP)))
  
  ## Remove monomer sites that overlap with coordinator and dimer site
  
  G_mon <- G_mon[ is.na(findOverlaps(G_mon, G_dimer, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  G_mon <- G_mon[ is.na(findOverlaps(G_mon, G_coord, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Monomer", length(G_mon)))
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Dimer", length(G_dimer)))
  
  G_Ebox <- G_Ebox[ is.na(findOverlaps(G_Ebox, G_coord, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  G_Ebox <- G_Ebox[ is.na(findOverlaps(G_Ebox, G_dimer, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  G_Ebox <- G_Ebox[ is.na(findOverlaps(G_Ebox, G_mon, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  
  EboxIndexDF <- G_Ebox %>%
    as.data.frame()
  
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Ebox", length(G_Ebox)))
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Coordinator", length(G_coord)))
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "TFAP", length(G_TFAP)))
  
  ## Plot count information
  SiteCounts <- SiteCounts %>%
    mutate(Count = as.numeric(Count)) %>%
    pivot_wider(names_from = "Desc", 
                values_from = "Count") %>%
    # mutate(MotifPositive = Initial - MNAsefilter) %>%
    mutate(Palindromes = MNAsefilter - NoPalindromes) %>%
    mutate(Overlap = NoPalindromes - NoOverlap, 
           CountableSites = MNAsefilter - Palindromes - Overlap) %>%
    select(-c(NoPalindromes, NoOverlap)) %>%
    pivot_longer(!Site, names_to = "Desc", values_to = "Count") 
  
  SiteCounts
  
  
  p1 <- SiteCounts %>%
    mutate(Desc = factor(Desc,
                         levels = c("Initial", "MNAsefilter", "Palindromes", "Overlap", "CountableSites"),
                         labels = c("Motif Found", "MNAse positive", "Palindrome", "Overlaps with\ncomposite site", "Bound Sites"), 
                         ordered = TRUE)) %>%
    ggplot(aes(x = Site,
               y = Count,
               label = Count, 
               fill = Desc)) +
    geom_col(position = "dodge", 
             alpha = 0.5, 
             size = 1) + 
    geom_text(size = 5,
              vjust = -1,
              color = "black",
              show.legend = FALSE,
              position = position_dodge(width = 0.9)) +
    ylab("Number of sites") +
    scale_fill_manual(name = "", 
                      values = viridis::viridis(5)) + 
    ylim(c(0, 5500)) +
    theme_bw() +
    theme(text = element_text(size = 20),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank(),
          axis.title.x = element_blank(),
          legend.position = "right")
  
  p2 <- SiteCounts %>% 
    filter(Desc == "CountableSites") %>%
    ggplot(aes(x = Site, y = Count, label = Count)) + 
    geom_col() + 
    geom_text(size = 6, 
              vjust = -0.5) + 
    theme_bw() + 
    theme(text = element_text(size = 24), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.text.x = element_text(angle = 45, 
                                     vjust = 1, 
                                     hjust = 1), 
          legend.position = "bottom")
  
  
  ## Concatenate all indices for filtering
  
  FilteredIndices <- c(G_dimer$name, 
                       G_mon$name, 
                       G_coord$name, 
                       G_Ebox$name,
                       G_TFAP$name)
  
  dfSignal_final_siteFiltered <- dfSignal_final %>% 
    filter(Motif_index %in% FilteredIndices)
  
  return(list(dfSignal_final_siteFiltered, p1, p2, dimerIndexDF, SiteCounts, EboxIndexDF))
  
}

```


```{r, fig.height = 4, fig.width = 8}

Output <- footprintFilter(dfSignal_final %>%                            
                            mutate(Signal = CR_signal))

dfSignal_final_siteFiltered <- Output[[1]]

dimerIndex <- Output[[4]]
EboxIndex <- Output[[6]]



for ( i in c(2:3)) {
  Output[i] %>% print()
}


```

### ALX4 CUT&RUN prior to filtering for digestion patterns

```{r, fig.width = 10, fig.height = 8}

footPrint(dfSignal_final  %>%
            mutate(CR = CR_signal)) 

```

### ALX4 CUT&RUN after filtering for digestion patterns

```{r, fig.width = 10, fig.height = 8}
footPrint(dfSignal_final_siteFiltered  %>%
            mutate(CR = CR_signal)) 

```


### TWIST1 CUT&RUN

```{r, fig.width = 10, fig.height = 8}

footPrint(dfSignal_final_siteFiltered  %>%
            mutate(CR = TWIST1_CR))

```

### ALX4 CUT&RUN in TWIST1 depleted conditions

```{r, fig.width = 10, fig.height = 8}
footPrint(dfSignal_final_siteFiltered  %>%
            mutate(CR = TWIST1_depleted)) 

```

### All plots knitted together

```{r, fig.width = 12, fig.height = 9}

dfSignal_final_siteFiltered %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  pivot_longer(c(CR_signal, TWIST1_depleted, TWIST1_CR), names_to = "Assay", values_to = "SignalValue") %>%
  group_by(MotifPosition, strand, name, Assay) %>%
  summarize(Avg = mean(SignalValue)) %>%
  ggplot(aes(x = MotifPosition, 
             y = Avg, 
             fill = strand)) +
  geom_col(show.legend = FALSE, 
           width = 1) + 
  facet_grid(Assay ~ name)+ 
  ylab("RPM") + 
  xlab("Distance from motif center (bp)") + 
  theme_bw() + 
  ylim(c(-0.05, 0.05)) + 
  scale_fill_manual(values = c("blue", "red")) + 
  scale_x_continuous(expand = c(0,3)) + 
  theme(text = element_text(size = 24), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.y = element_blank(), 
        axis.text.x = element_text(angle = 45, 
                                   hjust = 1, 
                                   vjust = 1),
        panel.margin = unit(1, "lines"), 
        strip.text.y = element_blank()
  )


```




## E-box independence

### Motif heatmap

```{r, fig.width = 3.2, fig.height = 5.6}

MotifHeatmap <- function(df, SiteTypeFilter) {
  
  a <- df %>% filter(SiteType == SiteTypeFilter) %>%
    filter(Motif_index %in% dfSignal_final_siteFiltered$Motif_index) %>%
    arrange(Motif.Score)
  
  range = 25
  
  Count <- dim(a) %>% .[1]
  
  print(Count)
  
  GRanges(seqnames = Rle(a$chr), 
          ranges = IRanges(start = a$MotifStart, 
                           end = a$MotifEnd), 
          strand = a$strand, 
          mcols = a %>% select(PeakSet)) %>%
    resize(., fix = "center", width = range * 2) %>%
    getSeq(hg19, .) %>%
    as.data.frame() %>%
    mutate(Motif = rownames(.)) %>%
    separate(col = x, into = c((-range - 1):range) %>% as.character(), sep = "") %>%
    select(-as.character(-range - 1)) %>%
    pivot_longer(!Motif, values_to = "Nucleotide", names_to = "Position") %>%
    mutate(Position = as.numeric(Position)) %>%
    ggplot(aes(x = Position, y = Motif, fill = Nucleotide)) + 
    geom_tile() + 
    scale_fill_manual(values = c("#DD0000", "#00BB00", "#0000EE", "#F8A200"), 
                      breaks = c("T", "A", "C", "G")) + 
    ylab(paste("MNAse confirmed ", SiteTypeFilter, "sites (n = ", Count, ")", sep = "")) + 
    scale_x_continuous(limits = c(-range, range), expand = c(0,0)) +
    theme(text = element_text(size = 18), 
          axis.text.y = element_blank(), 
          axis.title.y = element_text(size = 15),
          axis.ticks.y = element_blank(), 
          legend.position = "bottom")
  
}

MotifHeatmap(df, "Ebox") 

MotifHeatmap(df, "Coordinator")


```


### Independent site distances


```{r, fig.height = 4, fig.width = 4}

Ebox <- dfSignal_final_siteFiltered %>%
  filter(name == "Ebox") %>%
  distinct(Motif_index, .keep_all = TRUE)

Ebox_G <- 
  GRanges(seqnames = Rle(Ebox$seqnames), 
          ranges = IRanges(start = Ebox$mcols.MotifStart, 
                           end = Ebox$mcols.MotifEnd), 
          strand = Ebox$strand, 
          mcols = Ebox$Motif_index)


monomer <- dfSignal_final_siteFiltered %>%
  filter(name == "Monomer") %>%
  distinct(Motif_index, .keep_all = TRUE)


monomer_G <- 
  GRanges(seqnames = Rle(monomer$seqnames), 
          ranges = IRanges(start = monomer$mcols.MotifStart, 
                           end = monomer$mcols.MotifEnd), 
          strand = monomer$strand, 
          mcols = monomer$Motif_index)

independentSites <- distanceToNearest(Ebox_G, monomer_G, ignore.strand = TRUE) %>% 
  as.data.frame() %>% 
  filter(distance < 100) %>%
  merge(., monomer %>%
          mutate(rowNumber = row.names(.)) %>%
          select(mcols.MotifStart, mcols.MotifEnd, mcols.PeakName, rowNumber, strand), 
        by.x = "subjectHits", 
        by.y = "rowNumber") %>%
  rename(monomerStart = mcols.MotifStart, 
         monomerEnd = mcols.MotifEnd, 
         monomerPeak = mcols.PeakName, 
         monomerStrand = strand) %>%
  merge(., Ebox %>%
          mutate(rowNumber = row.names(.)) %>%
          select(mcols.MotifStart, mcols.MotifEnd, mcols.PeakName, rowNumber, strand), 
        by.x = "queryHits", 
        by.y = "rowNumber") %>%
  rename(EboxStart = mcols.MotifStart, 
         EboxEnd = mcols.MotifEnd, 
         EboxPeak = mcols.PeakName, 
         EboxStrand = strand) %>%
  mutate(distance = distance + 1) %>% ## +1 for flank of monomer site
  mutate(motif1 = case_when(monomerStart < EboxStart ~ "Monomer", 
                            TRUE ~ "Ebox")) %>%
  mutate(EboxStrand = as.character(EboxStrand), 
         monomerStrand = as.character(monomerStrand)) %>%
  mutate(distance = factor(distance, 
                           levels = c(0:100))) %>%
  filter( !( is.na(monomerStrand) )) %>%
  filter( !(is.na(EboxStrand))) %>%
  group_by(distance,
           .drop = FALSE) %>%
  summarize(count = n()) %>%
  mutate(distance = as.numeric(distance))

independentSites %>%
  ggplot(aes(x = distance, 
             y = count)) + 
  geom_line(size = 1) + 
  ggtitle("Closest distance between\nindependent HD monomer and E-box site") + 
  ylab("Number of motif pairs") + 
  xlab("Distance between motifs") + 
  scale_x_continuous(expand = c(0.005,0.005)) + 
  scale_y_continuous(expand = c(0.005,0.005)) + 
  theme_bw() + 
  theme(text = element_text(size = 24), 
        title = element_text(size = 13))


```

### Independent site distances - prior to site filter

```{r, fig.height = 4, fig.width = 4}

Ebox <- dfSignal_final %>%
  filter(name == "Ebox") %>%
  distinct(Motif_index, .keep_all = TRUE)

Ebox_G <- 
  GRanges(seqnames = Rle(Ebox$seqnames), 
          ranges = IRanges(start = Ebox$mcols.MotifStart, 
                           end = Ebox$mcols.MotifEnd), 
          strand = Ebox$strand, 
          mcols = Ebox$Motif_index)


monomer <- dfSignal_final %>%
  filter(name == "Monomer") %>%
  distinct(Motif_index, .keep_all = TRUE)


monomer_G <- 
  GRanges(seqnames = Rle(monomer$seqnames), 
          ranges = IRanges(start = monomer$mcols.MotifStart, 
                           end = monomer$mcols.MotifEnd), 
          strand = monomer$strand, 
          mcols = monomer$Motif_index)

allSites <- distanceToNearest(Ebox_G, monomer_G, ignore.strand = FALSE) %>% 
  as.data.frame() %>% 
  filter(distance < 100) %>%
  merge(., monomer %>%
          mutate(rowNumber = row.names(.)) %>%
          select(mcols.MotifStart, mcols.MotifEnd, mcols.PeakName, rowNumber, strand), 
        by.x = "subjectHits", 
        by.y = "rowNumber") %>%
  rename(monomerStart = mcols.MotifStart, 
         monomerEnd = mcols.MotifEnd, 
         monomerPeak = mcols.PeakName, 
         monomerStrand = strand) %>%
  merge(., Ebox %>%
          mutate(rowNumber = row.names(.)) %>%
          select(mcols.MotifStart, mcols.MotifEnd, mcols.PeakName, rowNumber, strand), 
        by.x = "queryHits", 
        by.y = "rowNumber") %>%
  rename(EboxStart = mcols.MotifStart, 
         EboxEnd = mcols.MotifEnd, 
         EboxPeak = mcols.PeakName, 
         EboxStrand = strand) %>%
  mutate(distance = distance + 1) %>% ## +1 for flank of monomer site
  mutate(motif1 = case_when(monomerStart < EboxStart ~ "Monomer", 
                            TRUE ~ "Ebox")) %>%
  mutate(EboxStrand = as.character(EboxStrand), 
         monomerStrand = as.character(monomerStrand)) %>%
  mutate(distance = factor(distance, 
                           levels = c(0:100))) %>%
  group_by(distance,
           .drop = FALSE) %>%
  summarize(count = n()) %>%
  mutate(distance = as.numeric(distance)) 

allSites %>%
  ggplot(aes(x = distance, 
             y = count)) + 
  geom_line(size = 1) + 
  ggtitle("Closest distance between\n any HD monomer and E-box site") + 
  ylab("Number of motif pairs") + 
  xlab("Distance between motifs") + 
  #facet_grid(monomerStrand ~ EboxStrand) + 
  scale_x_continuous(expand = c(0.005,0.005)) + 
  scale_y_continuous(expand = c(0.005,0.005)) + 
  theme_bw() + 
  theme(text = element_text(size = 24), 
        title = element_text(size = 13))


```

### Combined

```{r, fig.height = 5, fig.width = 11}

independentSites %>%
  mutate(Group = "Filtered independent E-box sites (n = 1026)") %>%
  rbind(allSites %>%
          mutate(Group = "All found E-box sites (n = 4552)")) %>% 
  ggplot(aes(x = distance, 
             y = count, 
             color = Group)) + 
  geom_line(size = 1) + 
  ggtitle("Closest distance between\nHD monomer and E-box site") + 
  ylab("Number of motif pairs") + 
  xlab("Distance between motifs") + 
  #facet_grid(monomerStrand ~ EboxStrand) + 
  scale_x_continuous(expand = c(0.005,0.005)) + 
  scale_y_continuous(expand = c(0.005,0.005)) + 
  theme_bw() + 
  theme(text = element_text(size = 24), 
        title = element_text(size = 20))



```



## Session info

```{r}

sessionInfo()

```