---
title: "Footprinting and post analysis"
output: 
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: false
    code_folding: hide
    theme: "paper"
    number_sections: true
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE, 
                      message = FALSE, 
                      error = TRUE, 
                      fig.width = 10) 
```

```{r}

library("tidyr")
library("DT")
library("cowplot")
library("tibble")
library("GenomicRanges")
library("biomaRt")
library("ggplot2")
library("plotly")
library("rtracklayer")
library("BSgenome.Hsapiens.UCSC.hg19")
library("rstatix")
library("ggpubr")
library("dplyr")
library("enrichR")
library("stringr")
library("scales")
library("grid")
library("gridExtra")

options(error=traceback)

```


## Organize initial Homer annotation

```{r}

## Read in called peak information
peaks <- read.delim(file = "All_ALX4_CR_peaks.bed", 
                header = FALSE)
colnames(peaks) <- c("chr", "start", "end", "name", "score", "strand")

peaks <- peaks %>%
  mutate(PeakSet = case_when(str_detect(name, "SRR24258480") ~ "TWISTposOnly",
                             TRUE ~ "Maintained"))

## Read in exact motif positions
bed <- read.delim(file = "All_ALX4_CR_peaks_motif.bed", 
                  header = FALSE, 
                  skip = 1)

colnames(bed) <- c("chr", "start", "end", "Motif", "MotifScore", "Strand")

bed <- mutate(bed, Motif = case_when(Motif == "1-CATCTGKTWTYAATTARN" ~ "Coordinator", 
                                     Motif == "5-TAATYGAATTAG" ~ "Dimer", 
                                     Motif == "2-YAATTA" ~ "Monomer", 
                                     Motif == "1-CAGATG" ~ "Ebox")) %>%
  arrange(desc(MotifScore))


## Remove mitochondrial DNA
peaks <- peaks %>%
  filter(chr != "chrM")

bed <- bed %>%
  filter(chr != "chrM")

```

## Determine number of possible sites per peak

```{r, fig.height = 5, fig.width = 5}

SiteCounts <- bed %>% group_by(Motif) %>% summarize(Count = n())
SiteCounts$Desc <- "Initial"

colnames(SiteCounts) <- c("Site", "Count", "Desc")

SiteCounts <- relocate(SiteCounts, "Desc", .before = "Site")

## All of my sites are off by 1 if they are on the negative strand, 
## I account for this here

G <- bed %>%
  mutate(start = as.numeric(start), 
         end = as.numeric(end)) %>%
  mutate(start = case_when(Strand == "-" ~ start + 1,
                           TRUE ~ start),
         end = case_when(Strand == "-" ~ end + 1,
                         TRUE ~ end))

G_Sites <- GRanges(seqnames = Rle(G$chr), 
          ranges = IRanges(start = G$start, 
                           end = G$end), 
          name = G$Motif,
          score = G$MotifScore,
          strand = G$Strand)


## Combine motif information with peak information

G_peaks <- GRanges(seqnames = Rle(peaks$chr), 
                   ranges = IRanges(start = peaks$start, 
                                    end = peaks$end),
                   names = peaks$name,
                   mcols = peaks %>% select(PeakSet))

OverlapIndices <- findOverlaps(G_Sites, G_peaks, 
             type = "any", 
             select = "all") %>%
  as.data.frame()

G$Motif.number <- rownames(G)
peaks$peak.number <- rownames(peaks)

df <- OverlapIndices %>% 
  merge(., G, by.x = "queryHits", by.y = "Motif.number", all.y = TRUE) %>%
  merge(., peaks, by.x = "subjectHits", by.y = "peak.number", all.y = TRUE) %>%
  select(-c(subjectHits, queryHits, chr.x, strand))
  
  
colnames(df) <- c("MotifStart", "MotifEnd", "SiteType", "Motif.Score", "strand", "chr", "PeakStart", "PeakEnd", "PeakName", "PeakScore", "PeakSet")
df$Motif_index <- rownames(df)

head(df)

hg19 <- BSgenome.Hsapiens.UCSC.hg19

```


# Motif heatmaps {.tabset}


## Function

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap <- function(df, SiteTypeFilter) {
  
  a <- df %>% filter(SiteType == SiteTypeFilter) %>%
    arrange(Motif.Score)
  
  range = 25
  
  Count <- SiteCounts %>%
    filter(Desc == "Initial") %>%
    filter(Site == SiteTypeFilter) %>%
    .$Count
  
  print(Count)
  
  plot1 <- GRanges(seqnames = Rle(a$chr), 
                   ranges = IRanges(start = a$MotifStart, 
                                    end = a$MotifEnd), 
                   strand = a$strand, 
                   mcols = a %>% select(PeakSet)) %>%
    resize(., fix = "center", width = range * 2) %>%
    getSeq(hg19, .) %>%
    as.data.frame() %>%
    mutate(Motif = rownames(.)) %>%
    separate(col = x, into = c((-range - 1):range) %>% as.character(), sep = "") %>%
    select(-as.character(-range - 1)) %>%
    pivot_longer(!Motif, values_to = "Nucleotide", names_to = "Position") %>%
    mutate(Position = as.numeric(Position)) %>%
    ggplot(aes(x = Position, y = Motif, fill = Nucleotide)) + 
    geom_tile() + 
    scale_fill_manual(values = c("#DD0000", "#00BB00", "#0000EE", "#F8A200"), 
                      breaks = c("T", "A", "C", "G")) + 
    ylab(paste(SiteTypeFilter, " \n(n = ", Count, ")", sep = "")) + 
    scale_x_continuous(limits = c(-range, range), expand = c(0,0)) +
    theme(text = element_text(size = 18), 
          axis.text.y = element_blank(), 
          axis.ticks.y = element_blank(), 
          legend.position = "bottom")
  
  plot2 <- GRanges(seqnames = Rle(a$chr), 
                   ranges = IRanges(start = a$MotifStart, 
                                    end = a$MotifEnd), 
                   strand = a$strand) %>%
    resize(., fix = "center", width = range * 2) %>%
    getSeq(hg19, .) %>%
    as.data.frame() %>%
    mutate(Motif = rownames(.)) %>%
    separate(col = x, into = c((-range-1):range) %>% as.character(), sep = "") %>%
    select(-as.character(-range - 1)) %>%
    pivot_longer(!Motif, values_to = "Nucleotide", names_to = "Position") %>%
    mutate(Position = as.numeric(Position)) %>%
    group_by(Position, Nucleotide) %>%
    summarise(count = n() / as.numeric(Count)) %>%
    ggplot(aes(x = Position, y = count, color = Nucleotide)) + 
    geom_line(linewidth = 1) + 
    scale_x_continuous(limits = c(-range, range), expand = c(0,0)) +
    scale_color_manual(values = c("#DD0000", "#00BB00", "#0000EE", "#F8A200"), 
                       breaks = c("T", "A", "C", "G")) + 
    ylab("Nucleotide\nfraction") + 
    theme_classic() + 
    theme(text = element_text(size = 18),
          legend.position = "none", 
          axis.title.x = element_blank(), 
          axis.ticks.x = element_blank(),
          axis.text.x = element_blank(),
          line = element_blank(),
          axis.ticks.y = element_blank())
  
  plot_grid(plot2, plot1, align = "v", rel_heights = c(2,6), nrow = 2)
  
}

```

## Dimer

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Dimer")

```


## Monomer

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Monomer")

```


## Coordinator

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Coordinator")

```

## Ebox

```{r, fig.width = 5, fig.height = 7.5}

MotifHeatmap(df, "Ebox")

```


These motifs only indicate whether a potential site is within a region of signal, not if that signal is actually bound. 

# Read in sequencing data

## Read in 5' read signal strength

Coverage files were generated using bedtools genomecov tool only from the <120bp regions for CUT&RUN footprinting and all reads for others.

See CoverageCreation.sh script for details. 


## Set regions of interest based on found sites

```{r}

bw_CR_plus <- read.delim("ALX4_CR/SRR24258480.rmdup_plus5_inPeaks.cov", header = FALSE)
bw_CR_plus$strand = "+"
bw_CR_minus <- read.delim("ALX4_CR/SRR24258480.rmdup_minus5_inPeaks.cov", header = FALSE)
bw_CR_minus$strand = "-"

## ASSUMING THAT THIS IS NOT RIGHT INCLUSIVE....

bw_CR <- rbind(bw_CR_plus, bw_CR_minus) %>%
  mutate(index = rownames(.)) %>%
  distinct() %>%
  mutate(V3 = V3 - 1)
colnames(bw_CR) <- c("chr", "start", "end", "depth", "strand", "index")

G_bw_CR <- GRanges(seqnames = Rle(bw_CR$chr),
                ranges = IRanges(start = bw_CR$start,
                                 end = bw_CR$end),
                strand = bw_CR$strand,
                mcols = bw_CR %>% select(depth))

```

## Grab signal strength at motif positions for CUT&RUN 5' ends

```{r}

df_onlySites <- df %>%
  filter( !(is.na(MotifStart)) )

G_Sites <- GRanges(Rle(df_onlySites$chr),
                   IRanges(start = df_onlySites$MotifStart,
                           end = df_onlySites$MotifEnd),
                   names = df_onlySites$Motif_index,
                   strand = df_onlySites$strand,
                   mcols = df_onlySites %>% select(SiteType, MotifStart, MotifEnd, PeakStart, PeakEnd, PeakName, PeakSet))

G_Sites_dispersed <- G_Sites %>%
  resize(width = 202, fix = "center") %>%
  as.data.frame() %>%
  rowwise() %>%
  mutate(binBreaks = ( seq(from = start, to = end, by = 1) %>% as.character() %>% paste(., collapse = ",")) ) %>%

  separate(binBreaks, into = c(1:202) %>% as.character(), sep = ",") %>%
  pivot_longer(!c(seqnames:mcols.PeakSet), names_to = "binPosition", values_to = "Coordinate") %>%
  filter( !( is.na(Coordinate) )) %>%
  mutate(Coordinate = as.numeric(Coordinate))

Overlaps <- findOverlaps(GRanges(seqnames = Rle(G_Sites_dispersed$seqnames),
                                IRanges(start = G_Sites_dispersed$Coordinate,
                                        end = G_Sites_dispersed$Coordinate),
                                names = G_Sites_dispersed$names,
                                strand = G_Sites_dispersed$strand,
                                mcols = G_Sites_dispersed %>% select(mcols.SiteType:mcols.PeakSet)),
                         G_bw_CR) %>%
  as.data.frame()

head(G_Sites_dispersed)

## Grab number of reads for normalization

totalReads <- read.delim(file = "ALX4_CR/SRR24258480_lt120.rmdup.bam_alignment_stats.txt",
                    skip = 6,
                    nrows = 1,
                    header = FALSE) %>%
  .$V3

dfSignal_final <-
  merge(Overlaps, G_Sites_dispersed %>%
                       as.data.frame() %>%
                       mutate(index = rownames(.)), by.x = "queryHits", by.y = "index") %>%
  merge(., G_bw_CR %>%
          as.data.frame() %>%
          mutate(index = rownames(.)), by.x = "subjectHits", by.y = "index") %>%
  rename(WindowStart = start.x,
         WindowEnd = end.x,
         MotifLength = width.x,
         strand = strand.x,
         seqnames = seqnames.x,
         binStart = start.y,
         binStop = end.y,
         binWidth = width.y,
         CR_signal = mcols.depth,
         PeakSet = mcols.PeakSet,
         name = mcols.SiteType) %>%
  select(-c(subjectHits, queryHits, seqnames.y, strand.y,
            binStart, binStop, binWidth, binPosition)) %>%
  mutate(MotifPosition = Coordinate - mcols.MotifStart - round( (mcols.MotifEnd - mcols.MotifStart) / 2),
    CR_signal = case_when(strand == "-" ~ -CR_signal,
                            TRUE ~ CR_signal)) %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  distinct() %>%
  mutate(CR_signal = CR_signal / ( totalReads / 1000000 ) )

## Ensure number every motif has each position accounted for

dfSignal_final %>%
  ggplot(aes(x = MotifPosition, fill = strand)) +
  geom_bar(width = 1) +
  facet_wrap(~name, nrow = 1) +
  ggtitle("CR signal read position distribution") +
  theme_bw() +
  theme(text = element_text(size = 24),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.y = element_blank(),
        axis.text.x = element_text(angle = 45, 
                                   hjust = 1, 
                                   vjust = 1),
        legend.position = "bottom")

## Read in chromosome sizes
chrom.sizes <- read.delim(file = "hg19.chrom.sizes.txt", 
                          header = FALSE)
colnames(chrom.sizes) <- c("seqnames", "chromSize")


## How many positions do not have full -100 to 100 coverage

print("Number of entries that do not have full coverage in the 100 bp window surrounding motif!")
dfSignal_final %>%
  merge(., chrom.sizes, by = "seqnames") %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  filter(mcols.MotifStart < 0 | mcols.MotifEnd > chromSize )

```
## Add relevant entries to matrix from other sequencing datasets

After each dataset is integrated, I will check to ensure that there are no entries that are NA in the dataset. This will show as empty datasets. 

### Function

```{r}
head(dfSignal_final)

AddtoSignalMatrix <- function(dfSignal,
                              seqType,
                              plusStrand,
                              minusStrand,
                              statsFile,
                              SKIP = 7){

  totalReads <- read.delim(file = statsFile,
                    skip = SKIP,
                    nrows = 1,
                    header = FALSE) %>%
  .$V3

  bw_plus <- read.delim(plusStrand, header = FALSE)
  bw_plus$strand = "+"
  bw_minus <- read.delim(minusStrand, header = FALSE)
  bw_minus$strand = "-"

  bw <- rbind(bw_plus, bw_minus) %>%
    distinct() %>%
    mutate(index = rownames(.)) %>%
    distinct() %>%
    mutate(V3 = V3 - 1)
  colnames(bw) <- c("chr", "start", "end", "depth", "strand", "index")


  G_bw <- GRanges(seqnames = Rle(bw$chr),
                ranges = IRanges(start = bw$start,
                                 end = bw$end),
                strand = bw$strand,
                mcols = bw %>% select(depth))


  tmp <- findOverlaps( GRanges(seqnames = Rle(dfSignal$seqnames),
                                        ranges = IRanges(start = dfSignal$Coordinate,
                                                         end = dfSignal$Coordinate),
                                        strand = dfSignal$strand,
                                        mcols = dfSignal %>% select(PeakSet)),
                                G_bw) %>%
    as.data.frame() %>%
    distinct() %>%
    merge(., dfSignal %>%
            mutate(index = rownames(.)), by.x = "queryHits", by.y = "index") %>%
    merge(., G_bw %>%
            as.data.frame() %>%
            mutate(index = rownames(.)), by.x = "subjectHits", by.y = "index") %>%
    rename(strand = strand.x,
           seqnames = seqnames.x,
           binStart = start,
           binStop = end,
           binWidth = width,
           NewDepth = mcols.depth) %>%
    select(-c(subjectHits, queryHits, seqnames.y, strand.y)) %>%
    mutate(NewDepth = case_when(strand == "-" ~ -NewDepth,
                                  TRUE ~ NewDepth)) %>%
    select(c(seqnames, Coordinate, NewDepth, strand)) %>%
    mutate(NewDepth = NewDepth / ( totalReads / 1000000 ) )


  dfSignal <- merge(dfSignal, tmp,
                          by = c("Coordinate", "seqnames", "strand"),
                          all.x = TRUE) %>%
    distinct()


  print("BELOW ENTRIES ARE NA AND NEED ADDRESSED!")
  dfSignal %>% filter(is.na(NewDepth)) %>% print()

  return(dfSignal)

}

```


## Add datasets to matrix
```{r}

## 5' reads TWIST1 depleted ALX4 CUT&RUN
dfSignal_final <- AddtoSignalMatrix(dfSignal_final,
                                    "CR_TWIST_depleted",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473.rmdup_plus5_inPeaks.cov",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473.rmdup_minus5_inPeaks.cov",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473_lt120.rmdup.bam_alignment_stats.txt",
                                    SKIP = 6) %>%
  rename(TWIST1_depleted = NewDepth)

# 5' reads TWIST1 C&R
dfSignal_final <- AddtoSignalMatrix(dfSignal_final,
                                    "TWIST1_CR",
                                    "TWIST1_CR/SRR24258481.rmdup_plus5_inPeaks.cov",
                                    "TWIST1_CR/SRR24258481.rmdup_minus5_inPeaks.cov",
                                   "TWIST1_CR/SRR24258481_lt120.rmdup.bam_alignment_stats.txt",
                                   SKIP = 6) %>%
  rename(TWIST1_CR = NewDepth)

```


## Initial footprint prior to filter {.tabset}

### Footprinting function

```{r, fig.height = 10, fig.width = 10}

dfSignal_final <- rename(dfSignal_final,
                         Motif_index = names) %>%
  mutate(MotifPosition = Coordinate -
           mcols.MotifStart - round( (mcols.MotifEnd - mcols.MotifStart) / 2))

MotifSortedbyStrength <- dfSignal_final %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  group_by(Motif_index, name) %>% 
  summarize(Avg_CR_sum = sum(CR_signal)) %>% 
  arrange(abs(Avg_CR_sum)) %>% 
  select(Motif_index)

footPrint <- function(dfSignal) {
  
  ID <- dfSignal %>%
    distinct(Motif_index, .keep_all = TRUE) %>%
    select(Motif_index, name) %>%
    mutate(Motif_index = factor(Motif_index, 
                                ordered = TRUE, 
                                #levels = MotifSortedbyStrength
                                )) %>%
    group_by(name) %>%
    mutate(ID = row_number())
  
  stichedPlot <- function(dfSignal, NAME) {
    
    if ( NAME == "Coordinator") {
      plot1 <- dfSignal %>%
        filter(name == NAME) %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        group_by(MotifPosition, strand, name) %>%
        summarize(Avg = mean(CR)) %>%
        ggplot(aes(x = MotifPosition, 
                   y = Avg, 
                   fill = strand)) +
        geom_col(show.legend = FALSE, 
                 width = 1) + 
        facet_wrap(~name, nrow = 1) + 
        ylab("RPM") + 
        theme_bw() + 
        ylim(c(-0.03, 0.03)) + 
        scale_fill_manual(values = c("blue", "red")) + 
        scale_x_continuous(expand = c(0,0)) + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.title.x = element_blank(), 
              axis.text.x = element_blank(), 
              axis.ticks.x = element_blank(),
              plot.margin = unit(c(8, 8, 0, 8), "points"), 
              panel.margin = unit(1, "lines"))
      
      plot2 <- dfSignal %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        filter(name == NAME) %>%
        group_by(MotifPosition, Motif_index, name) %>%
        summarize(CR = sum(CR %>% abs())) %>%
        merge(., ID, by = c("Motif_index", "name")) %>%
        ggplot(aes(x = MotifPosition, 
                   y = ID, 
                   fill = CR )) +
        geom_tile() + 
        scale_x_continuous(expand = c(0,0)) + 
        scale_y_discrete(expand = c(0,0)) + 
        scale_fill_gradient(low = "white", 
                            high = "blue", 
                            limits = c(0, .3), 
                            name = "RPM",
                            oob = scales::squish) +
        facet_wrap(~name, nrow = 1)+ 
        ylab("RPM") + 
        xlab("Distance from motif center (bp)") + 
        theme_bw() + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_blank(), 
              axis.ticks.y = element_blank(),
              axis.text.x = element_text(angle = 45, 
                                         hjust = 1, 
                                         vjust = 1), 
              axis.title.x = element_blank(),
              legend.position = "left",
              legend.key.width = unit(1, "cm"),
              legend.direction = "vertical", 
              plot.margin = unit(c(0, 8, 8, 8), "points"), 
              panel.margin = unit(1, "lines")
        )

    } else {
      
      plot1 <- dfSignal %>%
        filter(name == NAME) %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        group_by(MotifPosition, strand, name) %>%
        summarize(Avg = mean(CR)) %>%
        ggplot(aes(x = MotifPosition, 
                   y = Avg, 
                   fill = strand)) +
        geom_col(show.legend = FALSE, 
                 width = 1) + 
        facet_wrap(~name, nrow = 1) + 
        ylab("RPM") + 
        theme_bw() + 
        ylim(c(-0.03, 0.03)) + 
        scale_fill_manual(values = c("blue", "red")) + 
        scale_x_continuous(expand = c(0,0)) + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.text.y = element_blank(), 
              axis.ticks.y = element_blank(),
              axis.title.y = element_blank(), 
              axis.title.x = element_blank(), 
              axis.text.x = element_blank(), 
              axis.ticks.x = element_blank(),
              plot.margin = unit(c(8, 8, 0, 8), "points"), 
              panel.margin = unit(1, "lines"))
      
      plot2 <- dfSignal %>%
        filter(MotifPosition %in% c(-100:100)) %>%
        filter(name == NAME) %>%
        group_by(MotifPosition, Motif_index, name) %>%
        summarize(CR = sum(CR %>% abs())) %>%
        merge(., ID, by = c("Motif_index", "name")) %>%
        ggplot(aes(x = MotifPosition, 
                   y = ID, 
                   fill = CR )) +
        geom_tile(show.legend = FALSE) + 
        scale_x_continuous(expand = c(0,0)) + 
        scale_y_discrete(expand = c(0,0)) + 
        scale_fill_gradient(low = "white", 
                            high = "blue", 
                            limits = c(0, .3), 
                            name = "RPM", 
                            oob = scales::squish) +
        facet_wrap(~name, nrow = 1)+ 
        ylab("RPM") + 
        xlab("Distance from motif center (bp)") + 
        theme_bw() + 
        theme(text = element_text(size = 24), 
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_blank(), 
              axis.ticks.y = element_blank(),
              axis.text.x = element_text(angle = 45, 
                                         hjust = 1, 
                                         vjust = 1), 
              axis.title.x = element_blank(),
              plot.margin = unit(c(0, 8, 8, 8), "points"), 
              panel.margin = unit(1, "lines")
        )
    }
    
    plot2 <- plot2 + 
      theme(strip.background = element_blank(), strip.text = element_blank())

    plot_grid(plot1, plot2, 
              align = "v", 
              rel_heights = c(1,2.75), 
              nrow = 2, 
              axis = "lb")
    
  }
  
  p1 <- stichedPlot(dfSignal, "Coordinator")
  p2 <- stichedPlot(dfSignal, "Dimer")
  p3 <- stichedPlot(dfSignal, "Ebox")
  p4 <- stichedPlot(dfSignal, "Monomer")
  
  plot <- plot_grid(p1, p2, p3, p4,
                    align = "h", 
                    rel_widths = c(1.52,1,1,1), 
                    nrow = 1)
  
  x.grob <- textGrob("Distance from motif center (bp)", 
                     gp=gpar(fontsize=24))
  
  grid.arrange(arrangeGrob(plot, bottom = x.grob))
  
}


```

### ALX4 CUT&RUN

```{r, fig.height = 10, fig.width = 10}

footPrint(dfSignal_final %>%
  mutate(CR = CR_signal) )

```

### TWIST1 CUT&RUN

```{r, fig.height = 10, fig.width = 10}

footPrint(dfSignal_final %>%
  mutate(CR = TWIST1_CR))  

```


## Select sites based on footprinting

### Function

```{r}

footprintFilter <- function(dfSignal_final) {
  
  SiteCounts <- bed %>% group_by(Motif) %>% summarize(Count = n())
  SiteCounts$Desc <- "Initial"
  
  colnames(SiteCounts) <- c("Site", "Count", "Desc")
  
  SiteCounts <- relocate(SiteCounts, "Desc", .before = "Site")
  
  dimerIndex <- dfSignal_final %>% 
    filter(name == "Dimer") %>%
    mutate(Position = case_when(MotifPosition %in% c(-6:6) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-46:-7), c(7:46)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1)
  
  dimerIndexDF <- dfSignal_final %>% 
    filter(name == "Dimer") %>%
    mutate(Position = case_when(MotifPosition %in% c(-6:6) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-46:-7), c(7:46)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) )
  
  dimerIndex$Site <- "Dimer"
  
  monomerIndex <- dfSignal_final %>% 
    filter(name == "Monomer") %>%
    mutate(Position = case_when(MotifPosition %in% c(-3:3) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-43:-4), c(4:43)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1) 
  
  monomerIndex$Site <- "Monomer"
  
  coordinatorIndex <- dfSignal_final %>% 
    filter(name == "Coordinator") %>%
    mutate(Position = case_when(MotifPosition %in% c(-10:10) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-50:-11), c(11:50)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1) 
  
  coordinatorIndex$Site <- "Coordinator"
  
  EboxIndex <- dfSignal_final %>% 
    filter(name == "Ebox") %>%
    mutate(Position = case_when(MotifPosition %in% c(-3:3) ~ "MotifWindow", 
                                MotifPosition %in% c(c(-43:-4), c(4:43)) ~ "MotifFlank", 
                                TRUE ~ "Other")) %>%
    filter(Position %in% c("MotifWindow", "MotifFlank")) %>%
    group_by(Motif_index, Position) %>%
    summarize(Avg = mean(Signal %>% abs())) %>%
    pivot_wider(names_from = Position, 
                values_from = Avg) %>%
    mutate(Thres =  (MotifFlank + 0.01) / (MotifWindow + 0.01) ) %>%
    filter(Thres > 1)
  
  EboxIndex$Site <- "Ebox"
  
  SiteCounts <- rbind(SiteCounts, 
                      c("MNAsefilter", "Monomer", dim(monomerIndex)[1]),
                      c("MNAsefilter", "Dimer", dim(dimerIndex)[1]), 
                      c("MNAsefilter", "Coordinator", dim(coordinatorIndex)[1]), 
                      c("MNAsefilter", "Ebox", dim(EboxIndex)[1]))
  
  
  IndexDF <- rbind(dimerIndex, monomerIndex, coordinatorIndex, EboxIndex)
  
  
  tmp <- df %>% 
    distinct(Motif_index, .keep_all = TRUE) %>%
    merge(., IndexDF, by = "Motif_index") %>% 
    filter( !is.na(Site)) %>% 
    arrange(Thres %>% desc())
  
  Gtmp <- GRanges(seqnames = Rle(tmp$chr), 
                  ranges = IRanges(start = tmp$MotifStart, 
                                   end = tmp$MotifEnd), 
                  name = tmp$Motif_index,
                  strand = tmp$strand, 
                  mcols = tmp %>% select(MotifFlank:Site))
  
  G_Ebox <- Gtmp[mcols(Gtmp)$mcols.Site == "Ebox"]
  G_mon <- Gtmp[mcols(Gtmp)$mcols.Site == "Monomer"]
  G_dimer <- Gtmp[mcols(Gtmp)$mcols.Site == "Dimer"]
  G_coord <- Gtmp[mcols(Gtmp)$mcols.Site == "Coordinator"]
  
  
  ## Filter by self overlap
  
  G_mon <- G_mon[unique(findOverlaps(G_mon, 
                                     ignore.strand = TRUE,
                                     type = "any", 
                                     select = "first"))]
  

  
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Monomer", length(G_mon)))
  
  G_coord <- G_coord[unique(findOverlaps(G_coord,
                                         ignore.strand = TRUE, 
                                         type = "any", 
                                         select = "first"))]
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Coordinator", length(G_coord)))
  
  G_dimer <- G_dimer[unique(findOverlaps(G_dimer, 
                                         ignore.strand = TRUE,
                                         type = "any", 
                                         select = "first"))]
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Dimer", length(G_dimer)))
  
  G_Ebox <- G_Ebox[unique(findOverlaps(G_Ebox, 
                                       ignore.strand = TRUE,
                                       type = "any", 
                                       select = "first"))]
  SiteCounts <- rbind(SiteCounts, c("NoPalindromes", "Ebox", length(G_Ebox)))
  
  ## Remove monomer sites that overlap with coordinator and dimer site
  
  G_mon <- G_mon[ is.na(findOverlaps(G_mon, G_dimer, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  G_mon <- G_mon[ is.na(findOverlaps(G_mon, G_coord, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Monomer", length(G_mon)))
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Dimer", length(G_dimer)))
  
  G_Ebox <- G_Ebox[ is.na(findOverlaps(G_Ebox, G_coord, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  G_Ebox <- G_Ebox[ is.na(findOverlaps(G_Ebox, G_dimer, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  G_Ebox <- G_Ebox[ is.na(findOverlaps(G_Ebox, G_mon, ignore.strand = TRUE, type = "any", select = "arbitrary"))]
  
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Ebox", length(G_Ebox)))
  SiteCounts <- rbind(SiteCounts, c("NoOverlap", "Coordinator", length(G_coord)))
  
  ## Plot count information
  SiteCounts <- SiteCounts %>%
    mutate(Count = as.numeric(Count)) %>%
    pivot_wider(names_from = "Desc", 
                values_from = "Count") %>%
    # mutate(MotifPositive = Initial - MNAsefilter) %>%
    mutate(Palindromes = MNAsefilter - NoPalindromes) %>%
    mutate(Overlap = NoPalindromes - NoOverlap, 
           CountableSites = MNAsefilter - Palindromes - Overlap) %>%
    select(-c(NoPalindromes, NoOverlap)) %>%
    pivot_longer(!Site, names_to = "Desc", values_to = "Count") 
  
  SiteCounts
  
  
  p1 <- SiteCounts %>%
    mutate(Desc = factor(Desc,
           levels = c("Initial", "MNAsefilter", "Palindromes", "Overlap", "CountableSites"),
           labels = c("Motif Found", "MNAse positive", "Palindrome", "Overlaps with\ncomposite site", "Bound Sites"), 
           ordered = TRUE)) %>%
    ggplot(aes(x = Site,
               y = Count,
               label = Count, 
              fill = Desc)) +
    geom_col(position = "dodge", 
             alpha = 0.5, 
             size = 1) + 
    geom_text(size = 5,
              vjust = -1,
              color = "black",
              show.legend = FALSE,
              position = position_dodge(width = 0.9)) +
    ylab("Number of sites") +
    scale_fill_manual(name = "", 
                        values = viridis::viridis(5)) + 
    ylim(c(0, 5500)) +
    theme_bw() +
    theme(text = element_text(size = 20),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank(),
          axis.title.x = element_blank(),
          legend.position = "right")
  
  p2 <- SiteCounts %>% 
    filter(Desc == "CountableSites") %>%
    ggplot(aes(x = Site, y = Count, label = Count)) + 
    geom_col() + 
    geom_text(size = 6, 
              vjust = -0.5) + 
    theme_bw() + 
    theme(text = element_text(size = 24), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.text.x = element_text(angle = 45, 
                                     vjust = 1, 
                                     hjust = 1), 
          legend.position = "bottom")
  
  
  ## Concatenate all indices for filtering
  
  FilteredIndices <- c(G_dimer$name, 
                       G_mon$name, 
                       G_coord$name, 
                       G_Ebox$name)
  
  dfSignal_final_siteFiltered <- dfSignal_final %>% 
    filter(Motif_index %in% FilteredIndices)
  
  return(list(dfSignal_final_siteFiltered, p1, p2, dimerIndexDF, SiteCounts))
  
}

```


### Filter footprinting by ALX4 CUT&RUN {.tabset}

```{r, fig.height = 4, fig.width = 8}

Output <- footprintFilter(dfSignal_final %>%                            
                            mutate(Signal = CR_signal))

dfSignal_final_siteFiltered <- Output[[1]]

dimerIndex <- Output[[4]]



for ( i in c(2:3)) {
  Output[i] %>% print()
}


```

```{r, fig.width = 10, fig.height = 8}

footPrint(dfSignal_final  %>%
  mutate(CR = CR_signal))

footPrint(dfSignal_final_siteFiltered  %>%
  mutate(CR = CR_signal))

footPrint(dfSignal_final_siteFiltered  %>%
  mutate(CR = TWIST1_CR))

footPrint(dfSignal_final_siteFiltered  %>%
  mutate(CR = TWIST1_depleted))

```

```{r, fig.width = 10, fig.height = 9}

dfSignal_final_siteFiltered %>%
  filter(MotifPosition %in% c(-100:100)) %>%
  pivot_longer(c(CR_signal, TWIST1_depleted, TWIST1_CR), names_to = "Assay", values_to = "SignalValue") %>%
  group_by(MotifPosition, strand, name, Assay) %>%
  summarize(Avg = mean(SignalValue)) %>%
  ggplot(aes(x = MotifPosition, 
             y = Avg, 
             fill = strand)) +
  geom_col(show.legend = FALSE, 
           width = 1) + 
  facet_grid(Assay ~ name)+ 
  ylab("RPM") + 
  xlab("Distance from motif center (bp)") + 
  theme_bw() + 
  ylim(c(-0.03, 0.03)) + 
  scale_fill_manual(values = c("blue", "red")) + 
  scale_x_continuous(expand = c(0,3)) + 
  theme(text = element_text(size = 24), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.y = element_blank(), 
        axis.text.x = element_text(angle = 45, 
                                   hjust = 1, 
                                   vjust = 1),
        panel.margin = unit(1, "lines"), 
        strip.text.y = element_blank())


```


# Per peak basis

Make minor modifications to last dataframe generator function for per peak datasets

```{r}
AddtoSignalMatrix <- function(dfSignal,
                              seqType,
                              plusStrand,
                              minusStrand,
                              statsFile,
                              SKIP = 7){

  totalReads <- read.delim(file = statsFile,
                    skip = SKIP,
                    nrows = 1,
                    header = FALSE) %>%
  .$V3

  bw_plus <- read.delim(plusStrand, header = FALSE)
  bw_plus$strand = "+"
  bw_minus <- read.delim(minusStrand, header = FALSE)
  bw_minus$strand = "-"

  bw <- rbind(bw_plus, bw_minus) %>%
    distinct() %>%
    mutate(index = rownames(.)) %>%
    distinct() %>%
    mutate(V3 = V3 - 1)
  colnames(bw) <- c("chr", "start", "end", "depth", "strand", "index")


  G_bw <- GRanges(seqnames = Rle(bw$chr),
                ranges = IRanges(start = bw$start,
                                 end = bw$end),
                strand = bw$strand,
                mcols = bw %>% select(depth))


  tmp <- findOverlaps( GRanges(seqnames = Rle(dfSignal$seqnames),
                                        ranges = IRanges(start = dfSignal$Coordinate,
                                                         end = dfSignal$Coordinate),
                                        strand = dfSignal$strand,
                                        mcols = dfSignal %>% select(mcols.PeakSet)),
                                G_bw) %>%
    as.data.frame() %>%
    distinct() %>%
    merge(., dfSignal %>%
            mutate(index = rownames(.)), by.x = "queryHits", by.y = "index") %>%
    merge(., G_bw %>%
            as.data.frame() %>%
            mutate(index = rownames(.)), by.x = "subjectHits", by.y = "index") %>%
    rename(seqnames = seqnames.x,
           binStart = start.y,
           binStop = end.y,
           PeakStart = start.x, 
           PeakEnd = end.x,
           binWidth = width,
           NewDepth = mcols.depth, 
           strand = strand.x) %>%
    select(-c(subjectHits, queryHits, seqnames.y)) %>%
    select(c(seqnames, Coordinate, NewDepth, strand)) %>%
    mutate(NewDepth = NewDepth / ( totalReads / 1000000 ) )


  dfSignal <- merge(dfSignal, tmp,
                          by = c("Coordinate", "seqnames", "strand"),
                          all.x = TRUE) %>%
    distinct()


  print("BELOW ENTRIES ARE NA AND NEED ADDRESSED!")
  dfSignal %>% filter(is.na(NewDepth)) %>% print()

  return(dfSignal)

}

```


```{r, fig.height = 5, fig.width = 5}

BoundRegions <- dfSignal_final_siteFiltered %>% 
              distinct(Motif_index, seqnames, mcols.MotifStart, mcols.MotifEnd, name, PeakSet)


peaks <- peaks %>% 
  mutate(start = start + 1)

peaks$PeakIndex <- peaks$peak.number

dfBound <- df %>% 
  filter( (Motif_index %in% BoundRegions$Motif_index))

## Add peaks with no bound sites to dataframe
df_notBound <- df %>% 
  filter( !(PeakName %in% dfBound$PeakName )) %>%
  distinct(PeakName, .keep_all = TRUE) %>%
  mutate(MotifStart = NA, 
         MotifEnd = NA, 
         SiteType = NA, 
         Motif.Score = NA, 
         strand = NA)

dfBound <- rbind(dfBound, df_notBound)
dfBoundG <- GRanges(seqnames = Rle(dfBound$chr), 
                ranges = IRanges(start = dfBound$PeakStart, 
                                 end = dfBound$PeakEnd), 
                names = dfBound$PeakName,
                mcols = dfBound %>% select(-c(chr, PeakStart, PeakEnd)))


dfBound <- dfBoundG %>%
  resize(width = 200, fix = "center") %>%
  as.data.frame() %>%
  group_by(names, seqnames, start, end, mcols.SiteType, mcols.PeakSet) %>%
  summarize(Count = n()) %>%
  pivot_wider(names_from = "mcols.SiteType", 
              values_from = "Count")

dfBound_expand <- 
  dfBound %>% 
  rowwise() %>%
  mutate(binBreaks = ( seq(from = start, to = end, by = 1) %>% as.character() %>% paste(., collapse = ",")) ) %>%
  separate(binBreaks, into = c(1:200) %>% as.character(), sep = ",") %>%
  select(-`NA`) %>%
  pivot_longer(!c(names:Ebox), names_to = "binPosition", values_to = "Coordinate") %>%
  filter( !( is.na(Coordinate) )) %>%
  mutate(Coordinate = as.numeric(Coordinate)) %>%
  mutate(strand_1 = "+", 
         strand_2 = "-") %>%
  pivot_longer(c(strand_1, strand_2), names_to = "strand_tmp", values_to = "strand") %>%
  select(-strand_tmp)

dfBound_expand <- dfBound_expand %>%
  mutate(name = case_when( (Coordinator > 0 & is.na(Dimer) ) ~ "Coordinator", 
                           (Dimer > 0 & Coordinator > 0 ) ~ "Dimer + Coordinator", 
                           (Dimer > 0 & is.na(Coordinator) ) ~ "Dimer",
                           (Monomer > 0 & is.na(Ebox) ) ~ "Monomer", 
                           (Ebox > 0 & is.na(Monomer)) ~ "E-box",
                           (Monomer > 0 & Ebox > 0 ) ~ "Monomer + E-box",
                            TRUE ~ "No Site")) %>%
  mutate(name = factor(name, 
                       ordered = TRUE, 
                       levels = c("Coordinator", 
                                  "Dimer + Coordinator",
                                  "Dimer",
                                  "Monomer", 
                                  "E-box",
                                  "Monomer + E-box",
                                  "No Site")))

## Add sequencing information
## All reads CUT&RUN
dfBound_expand <- AddtoSignalMatrix(dfBound_expand,
                                    "CR_all",
                                    "ALX4_CR/SRR24258480.rmdup_plus_inPeaks.cov",
                                    "ALX4_CR/SRR24258480.rmdup_minus_inPeaks.cov",
                                    "ALX4_CR/SRR24258480.rmdup.bam_alignment_stats.txt") %>%
                  rename(CR_all = NewDepth)


## All reads TWIST1 depleted ALX4 CUT&RUN
dfBound_expand <- AddtoSignalMatrix(dfBound_expand,
                                    "CR_TWIST_depleted_all",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473.rmdup_plus_inPeaks.cov",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473.rmdup_minus_inPeaks.cov",
                                    "ALX4_CR_TWIST1_depleted/SRR24258473.rmdup.bam_alignment_stats.txt") %>%
                 rename(TWIST1_depleted_all = NewDepth)

## All reads TWIST1 depleted ALX4 CUT&RUN
dfBound_expand <- AddtoSignalMatrix(dfBound_expand,
                                    "ATAC",
                                    "ATAC/ATAC_merged_plus_inPeaks.cov",
                                    "ATAC/ATAC_merged_minus_inPeaks.cov",
                                    "ATAC/ATAC_merged.bam_alignment_stats.txt") %>%
                 rename(ATAC = NewDepth)

dfBound_expand <- AddtoSignalMatrix(dfBound_expand,
                                    "ATAC_TWIST1depleted",
                                    "TWIST1_depleted_ATAC/TWIST1_depleted_ATAC_plus_inPeaks.cov",
                                    "TWIST1_depleted_ATAC/TWIST1_depleted_ATAC_minus_inPeaks.cov",
                                    "TWIST1_depleted_ATAC/TWIST1_depleted_ATAC.bam_alignment_stats.txt") %>%
                 rename(ATAC_TWIST1depleted = NewDepth)

dfBound_expand %>%
  distinct(names, .keep_all = TRUE) %>%
  group_by(name) %>%
  summarize(Count = n()) %>%
  ggplot(aes(x = name, 
             y = Count, 
             label = Count, 
             fill = name, 
             color = name)) + 
  geom_col(show.legend = FALSE, 
           alpha = 0.3, 
           size = 0.7) +
  geom_text(vjust = -1, size = 5, color = "black") + 
  ggtitle("\nPeaks in each category") + 
    scale_colour_manual(values = c("#31688E", 
                                 "#35B779",
                                 "#440154", 
                                 "grey50", 
                                 "grey50", 
                                 "grey50", 
                                 "grey50")) + 
  scale_fill_manual(values = c(  "#31688E", 
                                 "#35B779", 
                                 "#440154", 
                                 "grey50", 
                                 "grey50", 
                                 "grey50", 
                                 "grey50")) + 
  ylim(c(0, 2600)) +
  theme_bw() +
  theme(text = element_text(size = 20),
        title = element_text(size = 14),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        legend.position = "bottom", 
        axis.title.x = element_blank(), 
        axis.text.x = element_text(angle = 30, 
                                   vjust = 1, 
                                   hjust = 1))

```

### Subtract reads per million between datasets

#### Compare RPMs across site types

```{r, fig.height = 5, fig.width = 5}

g <-  dfBound_expand %>%
  mutate(RPM_diff = abs(TWIST1_depleted_all) - abs(CR_all) ) %>%
  group_by(names, name) %>%
  summarize(Avg = mean(RPM_diff)) %>%
  ungroup() %>%
  wilcox_test(Avg ~ name) %>%
  add_xy_position()

g <- g %>% filter(group1 == "Dimer" | group2 == "Dimer") %>%
  mutate(xmin = case_when(group1 == "Dimer" ~ xmax,
                          TRUE ~ xmin)) %>%
  mutate(xmax = case_when(group2 == "Dimer" ~ xmin,
                          TRUE ~ xmax)) %>%
  mutate(p.adj = format(p.adj, scientific = TRUE, digits = 1))


g$y.position = rep(2, length(g$y.position))
g$xmin = g$xmax

dfBound_expand %>%
  mutate(RPM_diff = abs(TWIST1_depleted_all) -  abs(CR_all) ) %>%
  group_by(names, name) %>%
  summarize(Avg = mean(RPM_diff)) %>%
  ggplot(aes(x = name, y = Avg, 
             fill = name, 
             colour = name)) +
  geom_boxplot(alpha = 0.3, 
               show.legend = FALSE) +
  scale_colour_manual(values = c("#31688E", 
                                          "#35B779",
                                          "#440154", 
                                          "grey50", 
                                          "grey50", 
                                          "grey50", 
                                          "grey50")) + 
                                            scale_fill_manual(values = c(  "#31688E", 
                                                                                    "#35B779", 
                                                                                    "#440154", 
                                                                                    "grey50", 
                                                                                    "grey50", 
                                                                                    "grey50", 
                                                                                    "grey50")) + 
                                                                                      stat_pvalue_manual(g, 
                                                                                                         inherit.aes = FALSE,
                                                                                                         tip.length = 0, 
                                                                                                         size = 5, 
                                                                                                         label = "p.adj") +
  ylab("\n\u0394RPM") +
  ylim(c(-3, 2.5)) + 
  ggtitle("Difference in ALX4 signal in TWIST1-\nversus TWIST1+ backgrounds") + 
  theme_bw() +
  theme(text = element_text(size = 20),
        panel.grid.major = element_blank(),
        title = element_text(size = 14),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        legend.position = "bottom", 
        axis.title.x = element_blank(), 
        axis.text.x = element_text(angle = 30, 
                                   vjust = 1, 
                                   hjust = 1))
```

#### Compare RPMs across assays

```{r, fig.height = 6, fig.width = 12}


scalingFactors <- matrix(c("TWIST1_CR", "TWIST1_CR", 0.22, 96.04, 
                             "ALX4_CR", "CR_all", 0.12, 96.94, 
                             "ALX4_CR_TWIST-", "TWIST1_depleted_all", 0.10, 96.79), 
                         ncol = 4, byrow = TRUE) %>%
  as.data.frame()
                         
colnames(scalingFactors) <- c("Assay", "Identifier", "EcoliAlignment", "HumanAlignment")

normFactor <- (0.12 / 96.04) / (0.10 / 96.79 )

g <- dfBound_expand %>%
  mutate(TWIST1_depleted_all = TWIST1_depleted_all * normFactor) %>%
  pivot_longer(c(TWIST1_depleted_all, CR_all), names_to = "Assay", values_to = "Signal") %>%
  mutate(Assay = factor(Assay, 
                        labels = c("ALX4 C&R",
                                   "TWIST1- ALX4 C&R"), 
                        levels = c("CR_all", 
                                   "TWIST1_depleted_all"), 
                        ordered = TRUE)) %>%
  group_by(names, name, Assay) %>%
  summarize(Avg = mean(Signal %>% abs())) %>%
  ungroup() %>%
  group_by(name) %>%
  wilcox_test(Avg ~ Assay) %>%
  add_x_position() %>%
  mutate(y.position = 4.15) %>%
  mutate(p = case_when(name == "Coordinator" ~ "<2.2e-308", 
                       TRUE ~ as.character(p)))

dfBound_expand %>%
  mutate(TWIST1_depleted_all = TWIST1_depleted_all * normFactor) %>%
  pivot_longer(c(TWIST1_depleted_all, CR_all), names_to = "Assay", values_to = "Signal") %>%
  mutate(Assay = factor(Assay, 
                        labels = c("TWIST+",
                                   "TWIST1-"), 
                        levels = c("CR_all", 
                                   "TWIST1_depleted_all"), 
                        ordered = TRUE)) %>%
  group_by(names, name, Assay) %>%
  summarize(Avg = mean(Signal)) %>%
  group_by(name) %>%
  add_count() %>%
  mutate(n = n/2) %>%
  ggplot(aes(x = Assay, y = Avg, 
             fill = Assay, 
             colour = Assay)) +
  geom_boxplot(alpha = 0.3, 
               show.legend = FALSE, 
               size = 0.6) + 
  facet_wrap(~name, 
             nrow = 1,
             labeller = labeller(name = label_wrap_gen(width = 10))) +
  scale_colour_manual(values = c("#440154", 
                                "#35B779")) + 
  scale_fill_manual(values = c("#440154",
                               "#35B779")) + 
  stat_pvalue_manual(g, 
                     inherit.aes = FALSE,
                     tip.length = 0, 
                     size = 5, 
                     label = "p = {p}\n\n", 
                     bracket.size = 0.5) +
  geom_text(inherit.aes = FALSE, 
            x = 1.5, 
            y = 4.3, 
            size = 5, 
            check_overlap = TRUE,
            aes(label = paste("n = ", n, sep = ""))) +
  ylim(c(0,4.7)) +
  ylab("RPM") +
  ggtitle("ALX4 signal in TWIST1+ versus TWIST1- backgrounds") + 
  theme_bw() +
  theme(text = element_text(size = 20),
        panel.grid.major = element_blank(),
        title = element_text(size = 14),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        legend.position = "bottom", 
        axis.title.x = element_blank(), 
        axis.text.x = element_text(angle = 30, 
                                   vjust = 1, 
                                   hjust = 1))

```

## Accessibility changes across site types

```{r}

g <- dfBound_expand %>% 
  group_by(name, names) %>% 
  summarize(`TWIST+` = mean(ATAC), 
            `TWIST-` = mean(ATAC_TWIST1depleted)) %>% 
  ungroup() %>% 
  pivot_longer(c(`TWIST+`, `TWIST-`), 
               names_to = "ATAC", 
               values_to = "RPM") %>%
  mutate(ATAC = factor(ATAC, 
                       levels = c("TWIST+", "TWIST-"), 
                       labels = c("TWIST+", "TWIST-"), 
                       ordered = TRUE)) %>%
  group_by(name) %>%
  wilcox_test(RPM ~ ATAC) %>%
  add_xy_position() %>%
  mutate(y.position = 7.5)


dfBound_expand %>% 
  group_by(name, names) %>% 
  summarize(`TWIST+` = mean(ATAC), 
            `TWIST-` = mean(ATAC_TWIST1depleted)) %>% 
  ungroup() %>% 
  pivot_longer(c(`TWIST+`, `TWIST-`), 
               names_to = "ATAC", 
               values_to = "RPM") %>% 
    mutate(ATAC = factor(ATAC, 
                       levels = c("TWIST+", "TWIST-"), 
                       labels = c("TWIST+", "TWIST-"), 
                       ordered = TRUE)) %>%
  ggplot(aes(x = ATAC, 
             y = RPM, 
             fill = ATAC, 
             color = ATAC)) + 
  geom_boxplot(alpha = 0.3, 
               show.legend = FALSE,
               size = 0.6) + 
  stat_pvalue_manual(g, 
                     inherit.aes = FALSE,
                     tip.length = 0, 
                     size = 5, 
                     label = "p = {p}", 
                     bracket.size = 0.5) +
  ggtitle("Accessibility in TWIST1+ versus TWIST1- backgrounds") + 
  ylim(c(0,8)) + 
  facet_wrap(~name, 
             nrow = 1,
             labeller = labeller(name = label_wrap_gen(width = 10))) +
  scale_colour_manual(values = c("#440154", 
                                "#35B779")) + 
  scale_fill_manual(values = c("#440154",
                               "#35B779")) + 
  theme_bw() +
    theme(text = element_text(size = 20),
          panel.grid.major = element_blank(),
          title = element_text(size = 14),
          panel.grid.minor = element_blank(),
          panel.background = element_blank(),
          legend.position = "bottom", 
          axis.title.x = element_blank(), 
          axis.text.x = element_text(angle = 30, 
                                     vjust = 1, 
                                     hjust = 1))


```

## Motif composition per peak set

```{r, fig.height = 6, fig.width = 14}

peakCounts <- peaks %>% 
  group_by(PeakSet) %>% 
  summarise(PeakCount = n())

categoryCounts <- dfBound_expand %>%
  distinct(names, .keep_all = TRUE) %>%
  group_by(name) %>%
  summarize(Counts = n()) %>%
  mutate(Expected_TWIST1Indep = Counts * 0.1178462, 
         Expected_TWISTDep = Counts * 0.8821538)

g <- dfBound_expand %>% 
  distinct(names, .keep_all = TRUE) %>% 
  merge(., categoryCounts, by = "name") %>%
  group_by(name, mcols.PeakSet, Counts) %>%
  summarise(PeakNumber = n()) %>%
  pivot_wider(names_from = "mcols.PeakSet", 
              values_from = "PeakNumber") %>%
  merge(categoryCounts, 
        by = c("name", "Counts")) %>% 
  group_by(name) %>%
  summarise(chisq =  chisq_test(x = c(Maintained, TWISTposOnly), 
                           p = c(0.1178462, 0.8821538)) %>%
              .$p, 
            group1 = "TWIST1 dep", 
            group2 = "TWIST1 indep",
            xmin = 1, 
            xmax = 2,
            y.position = Counts + Counts*0.02)
    

dfBound_expand %>% 
  distinct(names, .keep_all = TRUE) %>% 
  merge(., categoryCounts, by = "name") %>%
  group_by(name, mcols.PeakSet, Counts) %>%
  summarise(PeakNumber = n()) %>%
  mutate(Observed = PeakNumber, 
         PercentageofPeaks = PeakNumber/Counts * 100, 
         Expected = case_when(mcols.PeakSet == "Maintained" ~ Counts * 0.1178462, 
                              mcols.PeakSet == "TWISTposOnly" ~ Counts * 0.8821538)) %>%
  pivot_longer(c(Observed, Expected), names_to = "OorE", values_to = "Value") %>%
  mutate(mcols.PeakSet = factor(mcols.PeakSet, 
                                levels = c("TWISTposOnly", "Maintained"), 
                                labels = c("TWIST1 dep", "TWIST1 indep"), 
                                ordered = TRUE)) %>%
  # filter(name %in% c("Coordinator", "Dimer", "Dimer + Coordinator")) %>%
  ggplot(aes(y = Value, 
             fill = OorE, 
             x = mcols.PeakSet)) + 
  geom_col(position = "dodge",
           alpha = 0.3, 
           size = 1) + 
    facet_wrap(~name, 
             nrow = 1, 
              scales = "free_y",  
             labeller = labeller(name = label_wrap_gen(width = 10))) + 
  stat_pvalue_manual(g, 
                     inherit.aes = FALSE,
                     tip.length = 0, 
                     size = 5, 
                     bracket.size = 0.5, 
                     label = "chisq") +
  ylab("\nNumber of peaks") + 
  scale_y_continuous(expand = c(0.05, 1)) + 
  ggtitle("Number of peaks that are TWIST1 dependent and independent\ncompared to expected") + 
  theme_bw() +
  theme(text = element_text(size = 20),
        panel.grid.major = element_blank(),
        title = element_text(size = 14),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        legend.position = "bottom", 
        axis.title.x = element_blank(), 
        axis.text.x = element_text(angle = 30, 
                                   vjust = 1, 
                                   hjust = 1))


dfBound_expand %>% 
  distinct(names, .keep_all = TRUE) %>% 
  merge(., categoryCounts, by = "name") %>%
  group_by(name, mcols.PeakSet, Counts) %>%
  summarise(PeakNumber = n()) %>%
  mutate(Observed = PeakNumber, 
         PercentageofPeaks = PeakNumber/Counts * 100, 
         Expected = case_when(mcols.PeakSet == "Maintained" ~ Counts * 0.1178462, 
                              mcols.PeakSet == "TWISTposOnly" ~ Counts * 0.8821538)) %>%
  pivot_longer(c(Observed, Expected), names_to = "OorE", values_to = "Value") %>%
  mutate(mcols.PeakSet = factor(mcols.PeakSet, 
                                levels = c("TWISTposOnly", "Maintained"), 
                                labels = c("TWIST1 dep", "TWIST1 indep"), 
                                ordered = TRUE)) %>%
  ggplot(aes(y = Value, 
             x = OorE, 
             fill = mcols.PeakSet)) + 
  geom_col(position = "stack",
           show.legend = FALSE,
           alpha = 0.3, 
           size = 1) + 
    facet_wrap(~name, 
             nrow = 1, 
              scales = "free_y",  
             labeller = labeller(name = label_wrap_gen(width = 10))) + 
  stat_pvalue_manual(g, 
                     inherit.aes = FALSE,
                     tip.length = 0, 
                     size = 5, 
                     bracket.size = 0.5, 
                     label = "chisq") +
  scale_fill_manual(name = "", 
                    labels = c("TWIST1 dependent", "TWIST1 independent"), 
                    values = viridis::viridis(2)) + 
  ylab("Number of peaks") + 
  scale_y_continuous(expand = expand_scale(mult = c(0.05, 0.1))) + 
  ggtitle("Number of peaks that are TWIST1 dependent vs independent compared to expected") + 
  theme_bw() +
  theme(text = element_text(size = 20),
        panel.grid.major = element_blank(),
        title = element_text(size = 14),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        legend.position = "bottom", 
        axis.title.x = element_blank(), 
        axis.text.x = element_text(angle = 30, 
                                   vjust = 1, 
                                   hjust = 1))


```
